### ISSUE TITLE
[Compatibility] Better checking of Forge compatibility

### ISSUE THREAD

**Issue by SquidDev** [sp614x/optifine#7422]
Each version of Optifine has a recommended version of Forge [listed on its download page](https://optifine.net/downloads). However, users often do not notice this, and so run Optifine with a different version of Forge than advertised.

While sometimes this is fine, often Optifine and Forge can conflict, leading to confusing crashes (i.e. https://github.com/sp614x/optifine/issues/7261, https://github.com/sp614x/optifine/issues/7159, https://github.com/sp614x/optifine/issues/7127). This is frustrating for both users (who have to deal with crashes) and mod authors (who often erroneously receive these bug reports, and have to triage these and offer tech support).

Ideally Optifine would be to warn about these issues much earlier on in the launch process, either displaying a warning screen when a different Forge version is installed, or preventing launching entirely.

**Comment by vinc3nz** [comment]
I totally agree. A warning screen or a chat message on server join would be good.

**Comment by sp614x** [comment]
Generally OptiFine supports a wide range of Forge versions around the compatible version.
Older as well as newer Forge versions can work just fine, depending on the changes that they add and if they are relevant to OF.

**Comment by SquidDev** [comment]
This is very much a case of most of the time it works, but when it doesn't it's a massive pain for everyone (for reference, I've received 16 crash reports caused by the above issues).

Forcing a specific Forge version would is definitely the most conservative option, but felt the easiest to implement. There are some other options I can think of:

 - Generate diffs against Forge's classes instead of vanilla's when running in an SRG environment, warning when those diffs do not correctly apply. This would allow for running on  Forge versions where patched classes have not changed.

 - After patching, ensure the exposed interface of the patched class is compatible with the original (Forge, not vanilla) class.

**Comment by sp614x** [comment]
Diffs against Forge classes are not practical as minimal factors (timestamps, compiler versions, cosmetic changes) can generate different bytecode for practically the same source. The MC classes are compiled once and sealed so diff against them is stable.
Checking if the patched class is backwards compatible with the base class is non-trivial and may generate false positives. In the best case it would also crash, but with error message.

I see that the reported issues are all about 1.19.2, it probably needs an update with the 1.19.4 patches.

**Comment by SquidDev** [comment]
> I see that the reported issues are all about OF 1.19.2, it probably needs an update with latest Forge patches.

If you've got time, that would be ideal - I've had yet another bug report today.

> Checking if the patched class is backwards compatible with the base class is non-trivial and may generate false positives. 

Even something basic like comparing the set of non-private methods+descriptors feels like it'd be pretty solid. What false positives are you thinking of here - if a method is added to the targetted class (even if it is a non-binary-change like overriding a method in a child class), Optifine is going to be incompatible in some (even if subtle) way.

**Comment by vinc3nz** [comment]
It is also possible to make a repository containing a list of all working (and non-working) forge versions. This list than can be updated by the community. If a  version is marked as not-working, the user can get a big warning screen saying that this forge version is known to cause issues

**Comment by sp614x** [comment]
There are many different kinds of incompatibilities and not all are immediately visible. 
From minor (exotic feature of some obscure mod not supported) to major (crash without other mods).
